internal/worker/accounts.go

```diff
@@ -0,0 +1,621 @@
+// called it will save all accounts.
+func NewAccountManager(key types.PrivateKey, owner string, alerter alerts.Alerter, w AccountMgrWorker, cs ConsensusState, dc DownloadContracts, s AccountStore, refillInterval time.Duration, l *zap.Logger) (*AccountMgr, error) {
+	logger := l.Named("accounts").Sugar()
+
+	shutdownCtx, shutdownCancel := context.WithCancel(context.Background())
+	a := &AccountMgr{
+		alerts: alerter,
+		w:      w,
+		cs:     cs,
+		dc:     dc,
+		s:      s,
+		key:    key,
+		logger: logger,
+		owner:  owner,
+
+		inProgressRefills:   make(map[types.PublicKey]struct{}),
+		lastLoggedRefillErr: make(map[types.PublicKey]time.Time),
+		refillInterval:      refillInterval,
+		shutdownCtx:         shutdownCtx,
+		shutdownCancel:      shutdownCancel,
+
+		byID: make(map[rhpv3.Account]*Account),
+	}
+	a.wg.Add(1)
+	go func() {
+		a.run()
+		a.wg.Done()
+	}()
+	return a, nil
+}
+
+// Account returns the account with the given id.
+func (a *AccountMgr) Account(hostKey types.PublicKey) api.Account {
+	acc := a.account(hostKey)
+	return acc.convert()
+}
+
+// Accounts returns all accounts.
+func (a *AccountMgr) Accounts() []api.Account {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	accounts := make([]api.Account, 0, len(a.byID))
+	for _, acc := range a.byID {
+		accounts = append(accounts, acc.convert())
+	}
+	return accounts
+}
+
+// ResetDrift resets the drift on an account.
+func (a *AccountMgr) ResetDrift(id rhpv3.Account) error {
+	a.mu.Lock()
+	account, exists := a.byID[id]
+	if !exists {
+		a.mu.Unlock()
+		return ErrAccountNotFound
+	}
+	a.mu.Unlock()
+
+	account.resetDrift()
+	return nil
+}
+
+func (a *AccountMgr) Shutdown(ctx context.Context) error {
+	accounts := a.Accounts()
+	err := a.s.UpdateAccounts(ctx, accounts)
+	if err != nil {
+		a.logger.Errorf("failed to save %v accounts: %v", len(accounts), err)
+		return err
+	}
+	a.logger.Infof("successfully saved %v accounts", len(accounts))
+
+	a.shutdownCancel()
+	a.wg.Wait()
+	return nil
+}
+
+func (a *AccountMgr) account(hk types.PublicKey) *Account {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	// Derive account key.
+	accKey := deriveAccountKey(a.key, hk)
+	accID := rhpv3.Account(accKey.PublicKey())
+
+	// Create account if it doesn't exist.
+	acc, exists := a.byID[accID]
+	if !exists {
+		acc = &Account{
+			key:    accKey,
+			logger: a.logger.Named(accID.String()),
+			acc: api.Account{
+				ID:            accID,
+				CleanShutdown: false,
+				HostKey:       hk,
+				Balance:       big.NewInt(0),
+				Drift:         big.NewInt(0),
+				Owner:         a.owner,
+				RequiresSync:  true, // force sync on new account
+			},
+		}
+		a.byID[accID] = acc
+	}
+	return acc
+}
+
+// ForHost returns an account to use for a given host. If the account
+// doesn't exist, a new one is created.
+func (a *AccountMgr) ForHost(hk types.PublicKey) *Account {
+	return a.account(hk)
+}
+
+func (a *AccountMgr) run() {
+	// wait for store to become available
```

Comment:
I'm not a huge fan of `worker.Setup` where we register the webhooks in a goroutine, but this feels very similar. Did you consider returning an account manager that's not running and move the initialisation to the worker's setup function?

Reply:
Yeah, but this just ended up being a lot simpler in terms of DX. Back when we assumed it's fine to just defer the setup until after the API is created it seemed like a great idea but if `Setup` just calls a function in the background that retries some operation a bunch of time I don't see the advantage anymore.